File: 10.py
#!/usr/bin/env python

from parse import parse


def parse_lights():
    with open('10.input') as f:
        for line in f:
            px, py, vx, vy = parse(
                'position=<{:6d}, {:6d}> velocity=<{:2d}, {:2d}>', line)
            yield (px, py), (vx, vy)


def bounds(frame):
    xmin = min(px for (px, _), _ in frame)
    xmax = max(px for (px, _), _ in frame)
    ymin = min(py for (_, py), _ in frame)
    ymax = max(py for (_, py), _ in frame)
    return (xmin, ymin), (xmax, ymax)


def size(bounds):
    (xmin, ymin), (xmax, ymax) = bounds
    return (xmax - xmin) * (ymax - ymin)


def step(frame):
    def step_one(light):
        (px, py), (vx, vy) = light
        return (px + vx, py + vy), (vx, vy)

    return set(step_one(l) for l in frame)


def play(frame):
    while True:
        frame = step(frame)
        yield frame


def congregate(start):
    gen, low, last = 0, size(bounds(start)), start
    for frame in play(start):
        sz = size(bounds(frame))
        if sz > low:
            return gen, low, last
        gen, low, last = gen + 1, sz, frame


def render(frame):
    points = set((px, py) for (px, py), _ in frame)

    def pixel(x, y):
        return '#' if (x, y) in points else ' '

    ((xmin, ymin), (xmax, ymax)) = bounds(frame)
    for y in range(ymin, ymax + 1):
        for x in range(xmin, xmax + 1):
            print(pixel(x, y), end='')
        print()


if __name__ == '__main__':
    start = set(parse_lights())
    gen, sz, frame = congregate(start)

    # part 1
    render(frame)

    # part 2
    print(gen)

File: 10.hs
#!/usr/bin/env stack
-- stack --resolver lts-12.10 script

{-# OPTIONS_GHC -Wall #-}

import Text.ParserCombinators.ReadP
import Utils

type Light = (Coord, Coord) -- (position, velocity)
type Frame = [Light]

coord' :: ReadP Coord
coord' = between (char '<') (char '>') coord

light :: ReadP Light
light = do
    position <- string  "position=" >> coord'
    velocity <- string " velocity=" >> coord'
    return (position, velocity)

step :: Frame -> Frame
step = map stepOne where
    stepOne ((px, py), (vx, vy)) = ((px + vx, py + vy), (vx, vy))

size :: Bounds -> Int
size ((xmin, ymin), (xmax, ymax)) = (xmax - xmin) * (ymax - ymin)

bounds :: Frame -> Bounds
bounds = boundingBox . map fst

play :: Frame -> [(Int, Frame)]
play = map (\frame -> (size (bounds frame), frame)) . iterate step

congregate :: Frame -> (Int, Frame, Int) -- return minimal (size, frame, gen)
congregate frame = firstMinima $ zipWith each frames generations where
    frames = play frame
    generations = iterate (+1) 0
    each (sz, f) gen = (sz, f, gen)

render :: Frame -> [String]
render f = go (bounds f) (map fst f) where
    go ((xmin, ymin), (xmax,ymax)) ps =
        [[pixel x y ps | x <- [xmin .. xmax]] | y <- [ymin .. ymax]]
    pixel x y ps = if (x, y) `elem` ps then '#' else '.'

main :: IO ()
main = do
    input <- readFile "10.input"
    let start = parseMany light $ lines input
    let (_, end, seconds) = congregate start
    -- part 1
    putStrLn $ unlines $ render end
    -- part 2
    print seconds
